File: typescript/problem_79.ts
export function canBeNonDecreasing(nums: number[]): boolean {
    const a = [...nums];
    let modified = false;
    for (let i = 1; i < a.length; i++) {
        if (a[i] < a[i-1]) {
            if (modified) return false;
            modified = true;
            if (i >= 2 && a[i] < a[i-2]) a[i] = a[i-1];
            else a[i-1] = a[i];
        }
    }
    return true;
}

export function problem79(): void {
    console.log("Problem 79:");
    if (!canBeNonDecreasing([10,5,7]))  throw new Error("assertion failed");
    if ( canBeNonDecreasing([10,5,1]))  throw new Error("assertion failed");
    if (!canBeNonDecreasing([1,2,3]))   throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: go/problem_79.go
package main

func canBeNonDecreasing(nums []int) bool {
	modified := false
	for i := 1; i < len(nums); i++ {
		if nums[i] < nums[i-1] {
			if modified {
				return false
			}
			modified = true
			if i >= 2 && nums[i] < nums[i-2] {
				nums[i] = nums[i-1]
			} else {
				nums[i-1] = nums[i]
			}
		}
	}
	return true
}

func problem79() {
	println("Problem 79:")
	if !canBeNonDecreasing([]int{10, 5, 7}) {
		panic("assertion failed")
	}
	if canBeNonDecreasing([]int{10, 5, 1}) {
		panic("assertion failed")
	}
	if !canBeNonDecreasing([]int{1, 2, 3}) {
		panic("assertion failed")
	}
	println("All Tests Passed")
}
